MapperService: check index.mapper.dynamic during index creation

davidvgalbraith opened the pull request:
The MapperService doesn't currently check the
index.mapper.dynamic setting during index creation,
so indices can be created with dynamic mappings even
if this setting is false. Add a check that throws an
exception in this case. Fixes #15381

jimferenczi commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingIntegrationIT.java:
```diff
@@ -32,6 +34,18 @@
 public class DynamicMappingIntegrationIT extends ESIntegTestCase {
 
+    public void testDynamicDisabled() throws IOException {
+        Settings settings = Settings.settingsBuilder().put("index.mapper.dynamic", false).build();
```
This setting is per index and it is honored every time (no bug). What's not working is when you set this setting for all indices.
To do this you can override nodeSettings in your test: 

    protected Settings nodeSettings(int nodeOrdinal) {
        return settingsBuilder().put(super.nodeSettings(nodeOrdinal))
            .put("index.mapper.dynamic", "false")
            .build();
    }

jimferenczi commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -318,6 +318,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                             continue;
                         }
                         try {
+                            mapperService.checkIfAddMappingAllowed(entry.getKey());
```
You cannot do the check here. You have no way to know if the type is created automatically by an index operation or through a legitimate create index.

jimferenczi commented:
@davidvgalbraith thank you for your PR.
I think you mixed the index.mapper.dynamic at the index level and at the root level. The bug is when you set index.mapper.dynamic.false for all indices (not for a single index). You should look at TransportIndexAction.java where the auto create is launched. I think it's here that you need to check if the index.mapper.dynamic should be honored or not (you can access the settings through the cluster state).

davidvgalbraith commented:
Thanks for the tips! I moved the check into `TransportIndexAction.java` as suggested. I couldn't get to the settings through the cluster state but conveniently they were already present in the `settings` variable.

jasontedor commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -88,6 +89,9 @@ protected void doExecute(final IndexRequest request, final ActionListener<IndexR
         // if we don't have a master, we don't have metadata, that's fine, let it find a master using create index API
         ClusterState state = clusterService.state();
         if (autoCreateIndex.shouldAutoCreate(request.index(), state)) {
+            if (!settings.getAsBoolean("index.mapper.dynamic", true)) {
```
Can we make `"index.mapper.dynamic"` a constant in `MapperService`?

jasontedor replied:
Also, I think that we should factor the default value out into a constant as well, and reuse it here and in `MapperService`.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledIT.java:
```diff
@@ -0,0 +1,29 @@
+import java.io.IOException;
+
+/**
+ * Created by davidgalbraith on 12/17/15.
```
We greatly appreciate the contribution, but we do not place attribution in our source code like this. :)

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledIT.java:
```diff
@@ -0,0 +1,29 @@
+            client().prepareIndex("index", "type", "1").setSource("foo", 3).get();
+            fail("Indexing request should have failed");
+        } catch (MapperParsingException e) {
+            // expected
```
Assert on the exception message, to prevent situations where the same exception is thrown but for a different reason.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledIT.java:
```diff
@@ -0,0 +1,29 @@
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder().put(super.nodeSettings(nodeOrdinal))
+            .put("index.mapper.dynamic", "false")
```
Use the factored out [constant](https://github.com/elastic/elasticsearch/pull/15424/files#r47975685) here.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledIT.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Created by davidgalbraith on 12/17/15.
+ */
+public class DynamicMappingDisabledIT extends ESIntegTestCase {
```
Is there any way to just make this a unit test (extending `ESTestCase`) instead of a full blown integration test?

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledIT.java:
```diff
@@ -0,0 +1,29 @@
+package org.elasticsearch.index.mapper;
```
If this [does end up needing to be an integration test](https://github.com/elastic/elasticsearch/pull/15424/files#r47975786), this file will need a license header.

rjernst commented:
I'm not sure we should be doing this change in general. We need to try to move away from defaults like this in the node configuration. It is error prone because one node missing the setting can cause weird behavior (eg some indexes not getting the setting depending on which nodes the index is created on).

Instead, the better path is to use index templates.

davidvgalbraith commented:
I replaced all occurrences of  `"index.mapper.dynamic"` with `MapperService.DYNAMIC_MAPPING_ENABLED_SETTING` and fixed the licensing. I kept the test as an integration test as I need to run a `TransportIndexAction` which requires a whole lot of infrastructure -- https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java#L73 that I don't have in a unit test scenario to my knowledge.

@rjernst I don't understand how index templates factor into this issue? The problem is that `TransportIndexAction` doesn't check its configured setting for `index.mapper.dynamic` before creating an index, so it'll still ignore it if it's in an index template.

rjernst commented:
@davidvgalbraith `TransportIndexAction` is looking at the node settings, which means it is looking for `index.mapper.dynamic` being set in `elasticsearch.yml` for that specific node the action runs on.

jimferenczi commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -88,6 +90,9 @@ protected void doExecute(final IndexRequest request, final ActionListener<IndexR
         ClusterState state = clusterService.state();
         if (autoCreateIndex.shouldAutoCreate(request.index(), state)) {
+            if (!settings.getAsBoolean(MapperService.DYNAMIC_MAPPING_ENABLED_SETTING, MapperService.DYNAMIC_MAPPING_ENABLED_DEFAULT)) {
+                throw new MapperParsingException("trying to auto create mapping, but dynamic mapping is disabled");
```
TypeMissingException ? It's a not a parsing exception per say. Can you add the index name and the type in the message.

davidvgalbraith commented:
@rjernst Yes! I agree that looking at the node settings is janky. However, it is the documented API, so I submit that we should make it work. We can improve the API in a subsequent version of Elasticsearch.

@jimferenczi I made it throw a `TypeMissingException` with a message including the index and type.

jasontedor commented on core/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -68,6 +68,8 @@
 public class MapperService extends AbstractIndexComponent implements Closeable {
 
     public static final String DEFAULT_MAPPING = "_default_";
+    public static final String DYNAMIC_MAPPING_ENABLED_SETTING = "index.mapper.dynamic";
```
Nit: Can you make this `INDEX_MAPPER_DYNAMIC_SETTING`?

jasontedor commented on core/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -68,6 +68,8 @@
 
     public static final String DEFAULT_MAPPING = "_default_";
+    public static final String DYNAMIC_MAPPING_ENABLED_SETTING = "index.mapper.dynamic";
+    public static final boolean DYNAMIC_MAPPING_ENABLED_DEFAULT = true;
```
Nit: Can you make this `INDEX_MAPPER_DYNAMIC_DEFAULT`?

davidvgalbraith replied:
I can! How did you come up with that name?

jasontedor replied:
The key used in settings is `index.mapper.dynamic`; I changed the `.` to `_` and made it uppercase.

jasontedor commented on core/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -120,7 +122,7 @@ public MapperService(IndexSettings indexSettings, AnalysisService analysisServic
         this.mapperRegistry = mapperRegistry;
 
-        this.dynamic = this.indexSettings.getSettings().getAsBoolean("index.mapper.dynamic", true);
+        this.dynamic = this.indexSettings.getSettings().getAsBoolean(DYNAMIC_MAPPING_ENABLED_SETTING, true);
```
This should be using the default value of the setting, instead of an explicit `true`.

davidvgalbraith commented:
@jasontedor I changed the variable names as suggested and used `INDEX_MAPPER_DYNAMIC_DEFAULT` where I had that `true`.

jasontedor commented:
> I kept the test as an integration test as I need to run a TransportIndexAction which requires a whole lot of infrastructure -- https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java#L73 that I don't have in a unit test scenario to my knowledge.

@davidvgalbraith Yes, `TransportIndexAction` requires a lot of infrastructure, but we have other `Transport.*Action` classes that have supporting unit tests (instead of integration tests) so it might be possible here too and that would be preferred over an integration test.

davidvgalbraith commented:
I made `DynamicMappingDisabledTests` a unit test.

davidvgalbraith commented:
Hey! Can we get a little more review on this / get it in?

jpountz commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -88,6 +90,10 @@ protected void doExecute(final IndexRequest request, final ActionListener<IndexR
+            if (!settings.getAsBoolean(MapperService.INDEX_MAPPER_DYNAMIC_SETTING, MapperService.INDEX_MAPPER_DYNAMIC_DEFAULT)) {
+                String message = String.format("trying to auto create mapping for [%s] in index [%s], but dynamic mapping is disabled", request.type(), request.index());
+                throw new TypeMissingException(null, request.type(), message);
+            }
```
Can you fold this logic into AutoCreateIndex so that it will also apply to delete/bulk/update requests?

jasontedor commented:
> Hey! Can we get a little more review on this / get it in?

@davidvgalbraith Please do note that activity is slow at this time of year and probably will be for another week due to seasonal holidays.

jasontedor commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -88,6 +90,10 @@ protected void doExecute(final IndexRequest request, final ActionListener<IndexR
         ClusterState state = clusterService.state();
         if (autoCreateIndex.shouldAutoCreate(request.index(), state)) {
+            if (!settings.getAsBoolean(MapperService.INDEX_MAPPER_DYNAMIC_SETTING, MapperService.INDEX_MAPPER_DYNAMIC_DEFAULT)) {
+                String message = String.format("trying to auto create mapping for [%s] in index [%s], but dynamic mapping is disabled", request.type(), request.index());
```
This is going to fail the forbidden APIs check for using the default locale; have you been running the `precommit` task or any tasks that depend on `precommit`?

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        threadPool = new ThreadPool("DynamicMappingDisabledTests");
```
You can get away with it right now because there is only one test, but this should be initialized once before the test suite, not once before each test in the suite.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+
+public class DynamicMappingDisabledTests extends ESSingleNodeTestCase {
+
+    private static ThreadPool threadPool;
```
Since this is static, the name should be `THREAD_POOL`.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    @AfterClass
+    public static void afterClass() {
```
Call this `destroyThreadPool`?

jasontedor commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -88,6 +90,10 @@ protected void doExecute(final IndexRequest request, final ActionListener<IndexR
         if (autoCreateIndex.shouldAutoCreate(request.index(), state)) {
+            if (!settings.getAsBoolean(MapperService.INDEX_MAPPER_DYNAMIC_SETTING, MapperService.INDEX_MAPPER_DYNAMIC_DEFAULT)) {
+                String message = String.format("trying to auto create mapping for [%s] in index [%s], but dynamic mapping is disabled", request.type(), request.index());
+                throw new TypeMissingException(null, request.type(), message);
```
Instead of passing `null` you can pass `new Index(request.index())`.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+
+            @Override
+            public void onFailure(Throwable e) {
+                assertEquals(e.getMessage(), "type[[type, trying to auto create mapping for [type] in index [index], but dynamic mapping is disabled]] missing");
```
Can we change this to just an `assertTrue` that `e.getMessage()` contains just the substring you added in `TransportIndexAction`?

jasontedor replied:
Should assert that `e` is an instance of `TypeMissingException`.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+        IndexRequest request = new IndexRequest("index", "type", "1");
+        request.source("foo", 3);
+
+        action.execute(request, new ActionListener<IndexResponse>() {
```
If neither `onFailure` nor `onResponse` are mistakenly not executed, this test will still pass. I think that we should add something to avoid that possibility.

jasontedor commented:
@davidvgalbraith I left some more review comments.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,100 @@
+        transportService = new TransportService(transport, threadPool);
+        indicesService = getInstanceFromNode(IndicesService.class);
+        shardStateAction = new ShardStateAction(Settings.EMPTY, clusterService, transportService, null, null);
+        actionFilters = new ActionFilters(new HashSet<>());
```
The `new HashSet<>()` can be replaced with `Collections.emptySet()`.

davidvgalbraith commented:
Thank you for the review today @jasontedor and @jpountz! I made all the suggested changes. Rolling the logic into `AutoCreateIndex` means it now throws an `IndexNotFoundException` when you try to autocreate an index with dynamic mappings disabled. That's the same thing it throws if automatic index creation is disabled, which seems like an admirable consistency.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,122 @@
+        @Override
+        public void onFailure(Throwable e) {
+            setOnFailureCalled(true);
+            assert(e instanceof IndexNotFoundException);
```
Can you make this a test assertion instead of a Java assertion?

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,122 @@
+    public void setUp() throws Exception {
+        super.setUp();
+        settings = Settings.builder()
+            .put(MapperService.INDEX_MAPPER_DYNAMIC_SETTING, "false")
```
You can just use the literal boolean `false` instead of the string `"false"`.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,122 @@
+
+        action.execute(request, listener);
+
+        assert(listener.getOnFailureCalled());
```
Let's also make this a test assertion instead of a Java assertion.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,122 @@
+    private AutoCreateIndex autoCreateIndex;
+    private Settings settings;
+
+    private class DynamicMappingDisabledTestActionListener implements ActionListener<IndexResponse> {
```
You can just make this an anonymous ActionListener as you had before, capturing an AtomicBoolean to flag whether or not `onFailure` was actually invoked. I think that it will clearer if it's inline with the test, and there will be a little less boilerplate because you can drop the field and the setter.

davidvgalbraith commented:
Thank you for the suggestions @jasontedor, I've implemented them all. Think we're getting close on this one!

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,112 @@
+            @Override
+            public void onFailure(Throwable e) {
+                onFailureCalled.set(true);
+                assertTrue(e instanceof IndexNotFoundException);
```
Sorry, what I meant in my previous comment about matchers is that this should be something like

```java
assertThat(e, instanceOf(IndexNotFoundException.class));
```

I'm going off-the-cuff here, you'll have to lookup the exact syntax and the appropriate matcher to statically import (`CoreMatchers#instanceOf`?). :)

This gives nicer error messages like "`Expected: an instance of org.elasticsearch.index.IndexNotFoundException but: ...`" instead of generic assertion errors.

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,112 @@
+
+        IndexRequest request = new IndexRequest("index", "type", "1");
+        request.source("foo", 3);
+        final AtomicBoolean onFailureCalled = new AtomicBoolean(false);
```
It defaults to `false`. :)

jasontedor commented on core/src/test/java/org/elasticsearch/index/mapper/DynamicMappingDisabledTests.java:
```diff
@@ -0,0 +1,112 @@
+    @AfterClass
+    public static void destroyThreadPool() {
+        ThreadPool.terminate(THREAD_POOL, 30, TimeUnit.SECONDS);
+        THREAD_POOL = null;
```
Let's leave a comment here about why `THREAD_POOL` is set to null (you can just steal the comment from other places like: `since static must set to null to be eligible for collection`). The collection is important since JVMs are reused during tests.

davidvgalbraith commented:
Ok, I changed the assertion to `assertThat(e, instanceOf(IndexNotFoundException.class));`, commented the nullification of `THREAD_POOL`, and used `new AtomicBoolean()` instead of `new AtomicBoolean(false)`.

jasontedor commented:
> Ok, I changed the assertion to `assertThat(e, instanceOf(IndexNotFoundException.class));`, commented the nullification of `THREAD_POOL`, and used `new AtomicBoolean()` instead of `new AtomicBoolean(false)`.

Excellent. Gradle check task passes locally for me so this is looking good. I'm going to defer to @jpountz for final sign-off and we can integrate after that. Please do note it is still a slow time of the year. :)

davidvgalbraith commented:
Hey thank you for the patience and guidance, @jasontedor!

jpountz commented:
This looks good to me too. I will merge later today.

jpountz commented:
I squashed and merged manually. Thanks @davidvgalbraith!
